File: cf_sdk/src/main/java/com/example/cf_sdk/logic/changebankapi/source/MemberRepository.java
package com.example.cf_sdk.logic.changebankapi.source;

import com.example.cf_sdk.logic.changebanksdk.model.FileResponse;
import com.example.cf_sdk.logic.changebanksdk.model.Session;
import com.example.cf_sdk.logic.changebanksdk.model.account.ProductDetailsReponseItem;
import com.example.cf_sdk.logic.changebanksdk.model.member.ActivateCardResponse;
import com.example.cf_sdk.logic.changebanksdk.model.member.CardDetailResponse;
import com.example.cf_sdk.logic.changebanksdk.model.member.DocumentApiResponse;
import com.example.cf_sdk.logic.changebanksdk.model.member.MemberDetails;
import com.example.cf_sdk.logic.changebanksdk.model.member.RSAPublicKeyResponse;
import com.example.cf_sdk.logic.changebanksdk.model.member.UserProfileResponse;
import com.example.cf_sdk.logic.changebanksdk.model.member.version.VersionConfig;
import com.example.cf_sdk.logic.changebanksdk.model.oow.OowQuestions;
import com.example.cf_sdk.logic.changebanksdk.parameter.account.CardToCardTransferParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.account.CheckDepositParam;
import com.example.cf_sdk.logic.changebanksdk.parameter.account.SubCardRelationParameter;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.ActivateCardParameter;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.AddAddressParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.AddConfidentialParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.AddMemberInfoAndAddressParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.AddProfilePictureParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.AgreementParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.CardDetailParameter;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.CreateBatchListMemberParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.CreateMemberParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.DocumentParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.ForgotPasswordParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.ForgotUsernameParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.IdscanParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.LogParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.MemberDetailParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.ProfilePictureParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.RefreshOowParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.ResetPasswordParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.SendPhoneCodeParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.SetESignAgreementAcceptedParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.SettingsParameter;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.UpdateEmailAddressParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.UploadDocumentsParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.UserProfileParameter;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.VerifyOutOfWalletParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.VerifyPhoneCodeParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.validation.EmailValidationParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.validation.PhoneValidationParameters;
import com.example.cf_sdk.logic.changebanksdk.response.ChangebankResponse;
import com.example.cf_sdk.logic.changebanksdk.response.DocumentUploadResponse;
import com.example.cf_sdk.logic.changebanksdk.response.IdscanResponse;
import com.example.cf_sdk.logic.changebanksdk.response.VerifyOowResponse;
import com.example.cf_sdk.logic.changebanksdk.source.MemberDatasource;
import com.google.common.base.Optional;

import java.io.File;
import java.net.HttpURLConnection;
import java.util.List;
import java.util.NoSuchElementException;

import javax.inject.Inject;
import javax.inject.Named;

import io.reactivex.Completable;
import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;

/**
 *
 * Member repository is in charge of choosing the correct {@link MemberDatasource}
 */

public class MemberRepository implements MemberDatasource {
    private MemberDatasource mRemoteDatasource;
    private MemberDatasource mCacheDatasource;

    private final Consumer<UserProfileResponse> saveUserProfile = new Consumer<UserProfileResponse>() {
        @Override
        public void accept(UserProfileResponse userProfileResponse) throws Exception {
            saveUserProfile(userProfileResponse);
        }
    };

    @Inject
    public MemberRepository(@Named("remote") MemberDatasource remoteDatasource,
                            @Named("cache") MemberDatasource cacheDatasource) {
        mRemoteDatasource = remoteDatasource;
        mCacheDatasource = cacheDatasource;
    }

    @Override
    public Single<ChangebankResponse> sendPhoneCode(
            SendPhoneCodeParameters sendPhoneCodeParameters) {
        return mRemoteDatasource.sendPhoneCode(sendPhoneCodeParameters);
    }

    @Override
    public Completable verifyPhoneCode(
            VerifyPhoneCodeParameters verifyPhoneCodeParameters) {
        return mRemoteDatasource.verifyPhoneCode(verifyPhoneCodeParameters);
    }

    @Override
    public Single<Session> createMember(CreateMemberParameters createMemberParameters) {
        return mRemoteDatasource.createMember(createMemberParameters);
    }

    @Override
    public Completable addAddress(final AddAddressParameters addAddressParameters) {
        return mRemoteDatasource.addAddress(addAddressParameters)
                .doOnComplete(new Action() {
                    @Override
                    public void run() throws Exception {
                        mCacheDatasource.addAddress(addAddressParameters);
                    }
                });
    }

    @Override
    public Single<MemberDetails> addMemberInfoAndAddress(final AddMemberInfoAndAddressParameters addMemberInfoAndAddressParameters) {
        return mRemoteDatasource.addMemberInfoAndAddress(addMemberInfoAndAddressParameters)
                .doOnSuccess(new Consumer<MemberDetails>() {
                    @Override
                    public void accept(MemberDetails memberDetails) throws Exception {
                        mCacheDatasource.saveMemberDetails(memberDetails);
                    }
                });
    }

    @Override
    public Single<Optional<OowQuestions>> addConfidential(final AddConfidentialParameters acp) {
  /*      Single<Optional<OowQuestions>> cache = mCacheDatasource.addConfidential(acp);

        Single<Optional<OowQuestions>> remote = mRemoteDatasource.addConfidential(acp)
                .doOnSuccess(new Consumer<Optional<OowQuestions>>() {
                    @Override
                    public void accept(Optional<OowQuestions> oowQuestionsOptional) throws Exception {
                        if (oowQuestionsOptional.isPresent()) {
                            mCacheDatasource.saveOowQuestions(oowQuestionsOptional.get());
                        }
                    }
                });

        return Single
                .concat(cache, remote)
                .filter(new Predicate<Optional<OowQuestions>>() {
                    @Override
                    public boolean test(Optional<OowQuestions> oowQuestionsOptional) throws Exception {
                        return oowQuestionsOptional.isPresent() && !oowQuestionsOptional.get().isExpired();
                    }
                })
                .firstOrError();*/

        return mRemoteDatasource.addConfidential(acp);
    }

    @Override
    public Completable saveOowQuestions(OowQuestions oowQuestions) {
        return mCacheDatasource.saveOowQuestions(oowQuestions);
    }

    @Override
    public Single<ChangebankResponse> phoneAvailability(PhoneValidationParameters pvp) {
        return mRemoteDatasource.phoneAvailability(pvp);
    }

    @Override
    public Single<ChangebankResponse> emailAvailability(EmailValidationParameters evp) {
        return mRemoteDatasource.emailAvailability(evp);
    }

    @Override
    public Single<MemberDetails> getCardToCardAccountDetails(CardToCardTransferParameters cardToCardTransferParameters) {
        return mRemoteDatasource.getCardToCardAccountDetails(cardToCardTransferParameters);
    }

    @Override
    public Single<MemberDetails> addCheckDeposit(CheckDepositParam checkDepositParam) {
        return mRemoteDatasource.addCheckDeposit(checkDepositParam);
    }

    @Override
    public Single<Optional<MemberDetails>> getMemberDetails(MemberDetailParameters parameters) {
        //Check cache first then remote
        Single<Optional<MemberDetails>> cache = mCacheDatasource.getMemberDetails(parameters);

        Single<Optional<MemberDetails>> remote = mRemoteDatasource.getMemberDetails(parameters)
                .doOnSuccess(new Consumer<Optional<MemberDetails>>() {
                    @Override
                    public void accept(Optional<MemberDetails> memberDetailsOptional) throws Exception {
                        if (memberDetailsOptional.isPresent()) {
                            saveMemberDetails(memberDetailsOptional.get());
                        }
                    }
                });

        return Single
                .concat(cache, remote)
                .filter(new Predicate<Optional<MemberDetails>>() {
                    @Override
                    public boolean test(Optional<MemberDetails> memberDetailsOptional) throws Exception {
                        return memberDetailsOptional.isPresent();
                    }
                })
                .firstOrError();
    }

    @Override
    public Single<VerifyOowResponse> verifyOowQuestion(VerifyOutOfWalletParameters verifyOutOfWalletParameters) {
        return mRemoteDatasource.verifyOowQuestion(verifyOutOfWalletParameters);
    }

    @Override
    public Single<OowQuestions> refreshOowQuestions(RefreshOowParameters refreshOowParameters) {
        return mRemoteDatasource.refreshOowQuestions(refreshOowParameters);
    }

    @Override
    public Single<File> getAgreement(AgreementParameters agreementParameters) {
        return mRemoteDatasource.getAgreement(agreementParameters);
    }


    @Override
    public void saveMemberDetails(MemberDetails memberDetails) {
        mCacheDatasource.saveMemberDetails(memberDetails);
    }

    @Override
    public void saveUserProfile(UserProfileResponse userProfileResponse) {
        mCacheDatasource.saveUserProfile(userProfileResponse);
    }

    @Override
    public void clearMemberDetailsCache() {
        mCacheDatasource.clearMemberDetailsCache();
    }

    @Override
    public void clearAllMemberDatasourceCache() {
        mCacheDatasource.clearAllMemberDatasourceCache();
    }

    @Override
    public Single<ChangebankResponse> setESignAgreementAsRead(SetESignAgreementAcceptedParameters setESignAgreementAcceptedParameters) {
        return mRemoteDatasource.setESignAgreementAsRead(setESignAgreementAcceptedParameters);
    }

    @Override
    public Single<Session> forgotPassword(ForgotPasswordParameters forgotPasswordParameters) {
        return mRemoteDatasource.forgotPassword(forgotPasswordParameters);
    }

    @Override
    public Single<ChangebankResponse> resetPassword(ResetPasswordParameters resetPasswordParameters) {
        return mRemoteDatasource.resetPassword(resetPasswordParameters);
    }

    @Override
    public Single<IdscanResponse> submitIdscan(IdscanParameters idscanParameters) {
        return mRemoteDatasource.submitIdscan(idscanParameters);
    }


    @Override
    public Single<FileResponse> uploadProfilePicture(final AddProfilePictureParameters addProfilePictureParameters) {
        return mRemoteDatasource.uploadProfilePicture(addProfilePictureParameters)
                .doOnSuccess(new Consumer<FileResponse>() {
                    @Override
                    public void accept(FileResponse fileResponse) throws Exception {
                        saveImageDocument(fileResponse);
                    }
                });
    }

    @Override
    public Single<ChangebankResponse> updateMemberPhone(final VerifyPhoneCodeParameters verifyPhoneCodeParameters) {
        return mRemoteDatasource.updateMemberPhone(verifyPhoneCodeParameters).onErrorResumeNext(new Function<Throwable, SingleSource<? extends ChangebankResponse>>() {
                    @Override
                    public SingleSource<? extends ChangebankResponse> apply(Throwable throwable) throws Exception {
                        if (throwable instanceof NoSuchElementException) {
                            ChangebankResponse error = new ChangebankResponse();
                            error.setHttpCode(HttpURLConnection.HTTP_NO_CONTENT);
                            error.setMessage("No Content");
                            mCacheDatasource.updateMemberPhone(verifyPhoneCodeParameters);
                            return Single.error(error);
                        }
                        return Single.error(throwable);
                    }
                })
                .doOnSuccess(new Consumer<ChangebankResponse>() {
                    @Override
                    public void accept(ChangebankResponse changebankResponse) throws Exception {
                        mCacheDatasource.updateMemberPhone(verifyPhoneCodeParameters);
                    }
                });
    }

    @Override
    public Single<ChangebankResponse> updateEmailAddress(final UpdateEmailAddressParameters updateEmailAddressParameters) {
        return mRemoteDatasource.updateEmailAddress(updateEmailAddressParameters).onErrorResumeNext(new Function<Throwable, SingleSource<? extends ChangebankResponse>>() {
                    @Override
                    public SingleSource<? extends ChangebankResponse> apply(Throwable throwable) throws Exception {

                        if (throwable instanceof NoSuchElementException) {
                            ChangebankResponse error = new ChangebankResponse();
                            error.setHttpCode(HttpURLConnection.HTTP_NO_CONTENT);
                            error.setMessage("No Content");
                            mCacheDatasource.updateEmailAddress(updateEmailAddressParameters);
                            return Single.error(error);
                        }
                        return Single.error(throwable);
                    }
                })
                .doOnSuccess(new Consumer<ChangebankResponse>() {
                    @Override
                    public void accept(ChangebankResponse changebankResponse) throws Exception {
                        mCacheDatasource.updateEmailAddress(updateEmailAddressParameters);
                    }
                });
    }

    @Override
    public Single<Optional<DocumentApiResponse>> getDocumentById(DocumentParameters documentParameters) {

        //Check cache first then remote
        Single<Optional<DocumentApiResponse>> cache = mCacheDatasource.getDocumentById(documentParameters);

        Single<Optional<DocumentApiResponse>> remote = mRemoteDatasource.getDocumentById(documentParameters)
                .doOnSuccess(new Consumer<Optional<DocumentApiResponse>>() {
                    @Override
                    public void accept(Optional<DocumentApiResponse> documentApiResponseOptional) throws Exception {
                        if (documentApiResponseOptional.isPresent()) {
                            saveDocument(documentApiResponseOptional.get());
                        }
                    }
                });

        return Single
                .concat(cache, remote)
                .filter(new Predicate<Optional<DocumentApiResponse>>() {
                    @Override
                    public boolean test(Optional<DocumentApiResponse> documentApiResponseOptional) throws Exception {
                        return documentApiResponseOptional.isPresent();
                    }
                })
                .firstOrError();
    }

    @Override
    public void saveDocument(DocumentApiResponse documentApiResponse) {
        mCacheDatasource.saveDocument(documentApiResponse);
    }

    @Override
    public Single<Optional<FileResponse>> getDocumentImageById(final DocumentParameters documentParameters) {
        //Check cache first then remote
        Single<Optional<FileResponse>> cache = mCacheDatasource.getDocumentImageById(documentParameters);

        Single<Optional<FileResponse>> remote = mRemoteDatasource.getDocumentImageById(documentParameters)
                .doOnSuccess(new Consumer<Optional<FileResponse>>() {
                    @Override
                    public void accept(Optional<FileResponse> fileResponseOptional) throws Exception {
                        if (fileResponseOptional.isPresent()) {
                            saveImageDocument(fileResponseOptional.get());
                        }
                    }
                });

        return Single
                .concat(cache, remote)
                .filter(new Predicate<Optional<FileResponse>>() {
                    @Override
                    public boolean test(Optional<FileResponse> fileResponseOptional) throws Exception {
                        return fileResponseOptional.isPresent();
                    }
                })
                .firstOrError();
    }

    @Override
    public Single<Optional<FileResponse>> getProfilePicture(ProfilePictureParameters parameters) {
        //Check cache first then remote
        Single<Optional<FileResponse>> cache = mCacheDatasource.getProfilePicture(parameters);

        Single<Optional<FileResponse>> remote = mRemoteDatasource.getProfilePicture(parameters)
                .doOnSuccess(new Consumer<Optional<FileResponse>>() {
                    @Override
                    public void accept(Optional<FileResponse> fileResponseOptional) throws Exception {
                        if (fileResponseOptional.isPresent()) {
                            saveImageDocument(fileResponseOptional.get());
                        }
                    }
                });

        return Single
                .concat(cache, remote)
                .filter(new Predicate<Optional<FileResponse>>() {
                    @Override
                    public boolean test(Optional<FileResponse> fileResponseOptional) throws Exception {
                        return fileResponseOptional.isPresent();
                    }
                })
                .firstOrError();
    }

    @Override
    public void saveImageDocument(FileResponse file) {
        mCacheDatasource.saveImageDocument(file);
    }

    @Override
    public Single<ChangebankResponse> log(LogParameters logParameters) {
        return mRemoteDatasource.log(logParameters);
    }

    @Override
    public Single<VersionConfig> getVersionConfig(SettingsParameter settingsParameter,String url) {
        return mRemoteDatasource.getVersionConfig(settingsParameter,url);
    }

    @Override
    public Single<UserProfileResponse> getUserProfile(UserProfileParameter parameters) {
        return mRemoteDatasource.getUserProfile(parameters).doOnSuccess(saveUserProfile);
    }

    @Override
    public Single<CardDetailResponse> getCardDetails(CardDetailParameter cardDetailParameter) {
        return mRemoteDatasource.getCardDetails(cardDetailParameter);
    }

    @Override
    public Single<List<ProductDetailsReponseItem>> getCardRelation(SubCardRelationParameter cardParameters) {
        return mRemoteDatasource.getCardRelation(cardParameters);
    }

    @Override
    public Single<RSAPublicKeyResponse> getRsaPublicKey(UserProfileParameter userProfileParameter) {
        return mRemoteDatasource.getRsaPublicKey(userProfileParameter);
    }

    @Override
    public Single<ActivateCardResponse> getActivateCard(ActivateCardParameter activateCardParameter) {
        return mRemoteDatasource.getActivateCard(activateCardParameter);
    }

    @Override
    public Completable forgotUsername(ForgotUsernameParameters forgotUsernameParameters,String appId) {
        return mRemoteDatasource.forgotUsername(forgotUsernameParameters,appId);
    }

    @Override
    public Completable batchListConfirm(CreateBatchListMemberParameters createBatchListMemberParameters) {
        return mRemoteDatasource.batchListConfirm(createBatchListMemberParameters);
    }

    @Override
    public Completable getAdminNumberAvailability(CreateBatchListMemberParameters createBatchListMemberParameters) {
        return mRemoteDatasource.getAdminNumberAvailability(createBatchListMemberParameters);
    }

    @Override
    public Single<Session> batchListCreateMember(CreateBatchListMemberParameters createBatchListMemberParameters) {
        return mRemoteDatasource.batchListCreateMember(createBatchListMemberParameters);
    }


    @Override
    public Completable deleteImageDocument(final ProfilePictureParameters documentParameters) {
        return mRemoteDatasource.deleteImageDocument(documentParameters)
                .doOnComplete(new Action() {
                    @Override
                    public void run() throws Exception {
                        mCacheDatasource.deleteImageDocument(documentParameters);
                    }
                });
    }


    // upload ID scan docs font,back selfie images
    @Override
    public Single<DocumentUploadResponse> uploadDocuments(final UploadDocumentsParameters idScanDocuments) {
        return mRemoteDatasource.uploadDocuments(idScanDocuments);

    }

}

File: cf_sdk/src/main/java/com/example/cf_sdk/changebankapi/source/MemberRepository.kt
package com.example.cf_sdk.changebankapi.source


import com.example.cf_sdk.changebankapi.model.FileResponse
import com.example.cf_sdk.defination.response.Session
import com.example.cf_sdk.changebankapi.model.account.ProductDetailsReponseItem
import com.example.cf_sdk.changebankapi.model.member.ActivateCardResponse
import com.example.cf_sdk.changebankapi.model.member.CardDetailResponse
import com.example.cf_sdk.changebankapi.model.member.DocumentApiResponse
import com.example.cf_sdk.changebankapi.model.member.MemberDetails
import com.example.cf_sdk.changebankapi.model.member.RSAPublicKeyResponse
import com.example.cf_sdk.defination.response.UserProfileResponse
import com.example.cf_sdk.defination.response.version.VersionConfig
import com.example.cf_sdk.changebankapi.model.oow.OowQuestions
import com.example.cf_sdk.changebankapi.parameter.account.CardToCardTransferParameters
import com.example.cf_sdk.changebankapi.parameter.account.CheckDepositParam
import com.example.cf_sdk.changebankapi.parameter.account.SubCardRelationParameter
import com.example.cf_sdk.changebankapi.parameter.member.ActivateCardParameter
import com.example.cf_sdk.changebankapi.parameter.member.AddAddressParameters
import com.example.cf_sdk.changebankapi.parameter.member.AddConfidentialParameters
import com.example.cf_sdk.changebankapi.parameter.member.AddMemberInfoAndAddressParameters
import com.example.cf_sdk.changebankapi.parameter.member.AddProfilePictureParameters
import com.example.cf_sdk.changebankapi.parameter.member.AgreementParameters
import com.example.cf_sdk.changebankapi.parameter.member.CardDetailParameter
import com.example.cf_sdk.changebankapi.parameter.member.CreateBatchListMemberParameters
import com.example.cf_sdk.changebankapi.parameter.member.CreateMemberParameters
import com.example.cf_sdk.changebankapi.parameter.member.DocumentParameters
import com.example.cf_sdk.changebankapi.parameter.member.ForgotPasswordParameters
import com.example.cf_sdk.changebankapi.parameter.member.ForgotUsernameParameters
import com.example.cf_sdk.changebankapi.parameter.member.IdscanParameters
import com.example.cf_sdk.changebankapi.parameter.member.LogParameters
import com.example.cf_sdk.changebankapi.parameter.member.MemberDetailParameters
import com.example.cf_sdk.changebankapi.parameter.member.ProfilePictureParameters
import com.example.cf_sdk.changebankapi.parameter.member.RefreshOowParameters
import com.example.cf_sdk.changebankapi.parameter.member.ResetPasswordParameters
import com.example.cf_sdk.changebankapi.parameter.member.SendPhoneCodeParameters
import com.example.cf_sdk.changebankapi.parameter.member.SetESignAgreementAcceptedParameters
import com.example.cf_sdk.defination.request.SettingsParameter
import com.example.cf_sdk.changebankapi.parameter.member.UpdateEmailAddressParameters
import com.example.cf_sdk.changebankapi.parameter.member.UploadDocumentsParameters
import com.example.cf_sdk.defination.request.UserProfileParameter
import com.example.cf_sdk.changebankapi.parameter.member.VerifyOutOfWalletParameters
import com.example.cf_sdk.changebankapi.parameter.member.VerifyPhoneCodeParameters
import com.example.cf_sdk.changebankapi.parameter.validation.EmailValidationParameters
import com.example.cf_sdk.changebankapi.parameter.validation.PhoneValidationParameters
import com.example.cf_sdk.defination.response.ChangebankResponse
import com.example.cf_sdk.changebankapi.response.DocumentUploadResponse
import com.example.cf_sdk.changebankapi.response.IdscanResponse
import com.example.cf_sdk.changebankapi.response.VerifyOowResponse
import com.example.cf_sdk.changebankapi.source.remote.MemberDatasource
import com.example.cf_sdk.defination.request.AccessTokenParameter
import com.example.cf_sdk.defination.request.AuthCodeParameter
import com.example.cf_sdk.defination.response.AuthCodeResponse

import com.google.common.base.Optional
import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.SingleSource
import io.reactivex.functions.Action
import io.reactivex.functions.Consumer
import io.reactivex.functions.Function
import java.io.File
import java.net.HttpURLConnection

/**
 *
 * Member repository is in charge of choosing the correct [MemberDatasource]
 */
class MemberRepository(
    private val mRemoteDatasource: MemberDatasource,
    private val mCacheDatasource: MemberDatasource
) : MemberDatasource {
    private val saveUserProfile: Consumer<UserProfileResponse> =
        Consumer { userProfileResponse -> saveUserProfile(userProfileResponse) }

    override fun sendPhoneCode(
        sendPhoneCodeParameters: SendPhoneCodeParameters?
    ): Single<ChangebankResponse?>? {
        return mRemoteDatasource.sendPhoneCode(sendPhoneCodeParameters)
    }

    override fun verifyPhoneCode(
        verifyPhoneCodeParameters: VerifyPhoneCodeParameters?
    ): Completable? {
        return mRemoteDatasource.verifyPhoneCode(verifyPhoneCodeParameters)
    }

    override fun createMember(createMemberParameters: CreateMemberParameters?): Single<Session?>? {
        return mRemoteDatasource.createMember(createMemberParameters)
    }

    override fun addAddress(addAddressParameters: AddAddressParameters?): Completable? {
        return mRemoteDatasource.addAddress(addAddressParameters)
            ?.doOnComplete(Action { mCacheDatasource.addAddress(addAddressParameters) })
    }

    override fun addMemberInfoAndAddress(addMemberInfoAndAddressParameters: AddMemberInfoAndAddressParameters?): Single<MemberDetails?>? {
        return mRemoteDatasource.addMemberInfoAndAddress(addMemberInfoAndAddressParameters)
            ?.doOnSuccess(Consumer { memberDetails ->
                mCacheDatasource.saveMemberDetails(
                    memberDetails
                )
            })
    }

    override fun addConfidential(acp: AddConfidentialParameters?): Single<Optional<OowQuestions?>?>? {
        /*      Single<Optional<OowQuestions>> cache = mCacheDatasource.addConfidential(acp);

        Single<Optional<OowQuestions>> remote = mRemoteDatasource.addConfidential(acp)
                .doOnSuccess(new Consumer<Optional<OowQuestions>>() {
                    @Override
                    public void accept(Optional<OowQuestions> oowQuestionsOptional) throws Exception {
                        if (oowQuestionsOptional.isPresent()) {
                            mCacheDatasource.saveOowQuestions(oowQuestionsOptional.get());
                        }
                    }
                });

        return Single
                .concat(cache, remote)
                .filter(new Predicate<Optional<OowQuestions>>() {
                    @Override
                    public boolean test(Optional<OowQuestions> oowQuestionsOptional) throws Exception {
                        return oowQuestionsOptional.isPresent() && !oowQuestionsOptional.get().isExpired();
                    }
                })
                .firstOrError();*/
        return mRemoteDatasource.addConfidential(acp)
    }

    override fun saveOowQuestions(oowQuestions: OowQuestions?): Completable? {
        return mCacheDatasource.saveOowQuestions(oowQuestions)
    }

    override fun phoneAvailability(pvp: PhoneValidationParameters?): Single<ChangebankResponse?>? {
        return mRemoteDatasource.phoneAvailability(pvp)
    }

    override fun emailAvailability(evp: EmailValidationParameters?): Single<ChangebankResponse?>? {
        return mRemoteDatasource.emailAvailability(evp)
    }

    override fun getCardToCardAccountDetails(cardToCardTransferParameters: CardToCardTransferParameters?): Single<MemberDetails?>? {
        return mRemoteDatasource.getCardToCardAccountDetails(cardToCardTransferParameters)
    }

    override fun addCheckDeposit(checkDepositParam: CheckDepositParam?): Single<MemberDetails?>? {
        return mRemoteDatasource.addCheckDeposit(checkDepositParam)
    }

    override fun getMemberDetails(parameters: MemberDetailParameters?): Single<Optional<MemberDetails?>?>? {
        //Check cache first then remote
        val cache = mCacheDatasource.getMemberDetails(parameters)
        val remote = mRemoteDatasource.getMemberDetails(parameters)
            ?.doOnSuccess(Consumer { memberDetailsOptional ->
                if (memberDetailsOptional?.isPresent == true) {
                    saveMemberDetails(memberDetailsOptional.get())
                }
            })
        return Single
            .concat(cache, remote)
            .filter { memberDetailsOptional -> memberDetailsOptional.isPresent }
            .firstOrError()
    }

    override fun verifyOowQuestion(verifyOutOfWalletParameters: VerifyOutOfWalletParameters?): Single<VerifyOowResponse?>? {
        return mRemoteDatasource.verifyOowQuestion(verifyOutOfWalletParameters)
    }

    override fun refreshOowQuestions(refreshOowParameters: RefreshOowParameters?): Single<OowQuestions?>? {
        return mRemoteDatasource.refreshOowQuestions(refreshOowParameters)
    }

    override fun getAgreement(agreementParameters: AgreementParameters?): Single<File?>? {
        return mRemoteDatasource.getAgreement(agreementParameters)
    }

    override fun saveMemberDetails(memberDetails: MemberDetails?) {
        mCacheDatasource.saveMemberDetails(memberDetails)
    }

    override fun saveUserProfile(userProfileResponse: UserProfileResponse?) {
        mCacheDatasource.saveUserProfile(userProfileResponse)
    }

    override fun clearMemberDetailsCache() {
        mCacheDatasource.clearMemberDetailsCache()
    }

    override fun clearAllMemberDatasourceCache() {
        mCacheDatasource.clearAllMemberDatasourceCache()
    }

    override fun setESignAgreementAsRead(setESignAgreementAcceptedParameters: SetESignAgreementAcceptedParameters?): Single<ChangebankResponse?>? {
        return mRemoteDatasource.setESignAgreementAsRead(setESignAgreementAcceptedParameters)
    }

    override fun forgotPassword(forgotPasswordParameters: ForgotPasswordParameters?): Single<Session?>? {
        return mRemoteDatasource.forgotPassword(forgotPasswordParameters)
    }

    override fun resetPassword(resetPasswordParameters: ResetPasswordParameters?): Single<ChangebankResponse?>? {
        return mRemoteDatasource.resetPassword(resetPasswordParameters)
    }

    override fun submitIdscan(idscanParameters: IdscanParameters?): Single<IdscanResponse?>? {
        return mRemoteDatasource.submitIdscan(idscanParameters)
    }

    override fun uploadProfilePicture(addProfilePictureParameters: AddProfilePictureParameters?): Single<FileResponse?>? {
        return mRemoteDatasource.uploadProfilePicture(addProfilePictureParameters)
            ?.doOnSuccess(Consumer { fileResponse -> saveImageDocument(fileResponse) })
    }

    override fun updateMemberPhone(verifyPhoneCodeParameters: VerifyPhoneCodeParameters?): Single<ChangebankResponse?>? {
        return mRemoteDatasource.updateMemberPhone(verifyPhoneCodeParameters)!!
            .onErrorResumeNext(object : Function<Throwable?, SingleSource<out ChangebankResponse>> {
                @Throws(Exception::class)


                override fun apply(t: Throwable): SingleSource<out ChangebankResponse> {
                    if (t is NoSuchElementException) {
                        val error =
                            ChangebankResponse()
                        error.httpCode = HttpURLConnection.HTTP_NO_CONTENT
                        error.setMessage("No Content")
                        mCacheDatasource.updateMemberPhone(verifyPhoneCodeParameters)
                        return Single.error(error)
                    }
                    return Single.error(t)
                }
            })
            .doOnSuccess { mCacheDatasource.updateMemberPhone(verifyPhoneCodeParameters) }
    }

    override fun updateEmailAddress(updateEmailAddressParameters: UpdateEmailAddressParameters?): Single<ChangebankResponse?>? {
        return mRemoteDatasource.updateEmailAddress(updateEmailAddressParameters)!!
            .onErrorResumeNext(object : Function<Throwable?, SingleSource<out ChangebankResponse>> {
                @Throws(Exception::class)
                override fun apply(t: Throwable): SingleSource<out ChangebankResponse> {
                    if (t is NoSuchElementException) {
                        val error =
                            ChangebankResponse()
                        error.httpCode = HttpURLConnection.HTTP_NO_CONTENT
                        error.setMessage("No Content")
                        mCacheDatasource.updateEmailAddress(updateEmailAddressParameters)
                        return Single.error(error)
                    }
                    return Single.error(t)
                }
            })
            .doOnSuccess { mCacheDatasource.updateEmailAddress(updateEmailAddressParameters) }
    }

    override fun getDocumentById(documentParameters: DocumentParameters?): Single<Optional<DocumentApiResponse?>?>? {

        //Check cache first then remote
        val cache = mCacheDatasource.getDocumentById(documentParameters)
        val remote = mRemoteDatasource.getDocumentById(documentParameters)
            ?.doOnSuccess(Consumer { documentApiResponseOptional ->
                if (documentApiResponseOptional?.isPresent == true) {
                    saveDocument(documentApiResponseOptional.get())
                }
            })
        return Single
            .concat(cache, remote)
            .filter { documentApiResponseOptional -> documentApiResponseOptional.isPresent }
            .firstOrError()
    }

    override fun saveDocument(documentApiResponse: DocumentApiResponse?) {
        mCacheDatasource.saveDocument(documentApiResponse)
    }

    override fun getDocumentImageById(documentParameters: DocumentParameters?): Single<out Optional<out FileResponse?>?> {
        //Check cache first then remote
        val cache = mCacheDatasource.getDocumentImageById(documentParameters)
        val remote = mRemoteDatasource.getDocumentImageById(documentParameters)
            ?.doOnSuccess(Consumer { fileResponseOptional ->
                if (fileResponseOptional?.isPresent == true) {
                    saveImageDocument(fileResponseOptional.get())
                }
            })
        return Single
            .concat(cache, remote)
            .filter { fileResponseOptional -> fileResponseOptional.isPresent }
            .firstOrError()
    }

    override fun getProfilePicture(parameters: ProfilePictureParameters?): Single<out Optional<out FileResponse?>?> {
        //Check cache first then remote
        val cache = mCacheDatasource.getProfilePicture(parameters)
        val remote = mRemoteDatasource.getProfilePicture(parameters)
            ?.doOnSuccess(Consumer { fileResponseOptional ->
                if (fileResponseOptional?.isPresent == true) {
                    saveImageDocument(fileResponseOptional.get())
                }
            })
        return Single
            .concat(cache, remote)
            .filter { fileResponseOptional -> fileResponseOptional.isPresent }
            .firstOrError()
    }

    override fun saveImageDocument(file: FileResponse?) {
        mCacheDatasource.saveImageDocument(file)
    }

    override fun log(logParameters: LogParameters?): Single<ChangebankResponse?>? {
        return mRemoteDatasource.log(logParameters)
    }

    override fun getVersionConfig(settingsParameter: SettingsParameter?): Single<VersionConfig?>? {
        return mRemoteDatasource.getVersionConfig(settingsParameter)
    }

    override fun getAuthCode(authCodeParameter: AuthCodeParameter?): Single<AuthCodeResponse?>? {
        return mRemoteDatasource.getAuthCode(authCodeParameter)
    }


    override fun getUserProfile(parameters: UserProfileParameter?): Single<UserProfileResponse?>? {
        return mRemoteDatasource.getUserProfile(parameters)
    }

    override fun getCardDetails(cardDetailParameter: CardDetailParameter?): Single<CardDetailResponse?>? {
        return mRemoteDatasource.getCardDetails(cardDetailParameter)
    }

    override fun getCardRelation(cardParameters: SubCardRelationParameter?): Single<List<ProductDetailsReponseItem?>?>? {
        return mRemoteDatasource.getCardRelation(cardParameters)
    }

    override fun getRsaPublicKey(userProfileParameter: UserProfileParameter?): Single<RSAPublicKeyResponse?>? {
        return mRemoteDatasource.getRsaPublicKey(userProfileParameter)
    }

    override fun getActivateCard(activateCardParameter: ActivateCardParameter?): Single<ActivateCardResponse?>? {
        return mRemoteDatasource.getActivateCard(activateCardParameter)
    }

    override fun forgotUsername(
        forgotUsernameParameters: ForgotUsernameParameters?,
        appId: String?
    ): Completable? {
        return mRemoteDatasource.forgotUsername(forgotUsernameParameters, appId)
    }

    override fun batchListConfirm(createBatchListMemberParameters: CreateBatchListMemberParameters?): Completable? {
        return mRemoteDatasource.batchListConfirm(createBatchListMemberParameters)
    }

    override fun getAdminNumberAvailability(createBatchListMemberParameters: CreateBatchListMemberParameters?): Completable? {
        return mRemoteDatasource.getAdminNumberAvailability(createBatchListMemberParameters)
    }

    override fun batchListCreateMember(createBatchListMemberParameters: CreateBatchListMemberParameters?): Single<Session?>? {
        return mRemoteDatasource.batchListCreateMember(createBatchListMemberParameters)
    }

    override fun deleteImageDocument(documentParameters: ProfilePictureParameters?): Completable? {
        return mRemoteDatasource.deleteImageDocument(documentParameters)
            ?.doOnComplete(Action { mCacheDatasource.deleteImageDocument(documentParameters) })
    }

    // upload ID scan docs font,back selfie images
    override fun uploadDocuments(idScanDocuments: UploadDocumentsParameters?): Single<DocumentUploadResponse?>? {
        return mRemoteDatasource.uploadDocuments(idScanDocuments)
    }
}